Tests/helper.py

```diff
@@ -250,25 +251,27 @@ def tostring(im: Image.Image, string_format: str, **options: Any) -> bytes:
     return out.getvalue()
 
 
-def hopper(mode: str | None = None, cache: dict[str, Image.Image] = {}) -> Image.Image:
+def hopper(mode: str | None = None) -> Image.Image:
+    # Use caching to reduce reading from disk but so an original copy is
+    # returned each time and the cached image isn't modified by tests
+    # (for fast, isolated, repeatable tests).
+    return _cached_hopper(mode).copy()
+
+
+@lru_cache(maxsize=None)
+def _cached_hopper(mode: str | None = None) -> Image.Image:
     if mode is None:
         # Always return fresh not-yet-loaded version of image.
         # Operations on not-yet-loaded images is separate class of errors
         # what we should catch.
         return Image.open("Tests/images/hopper.ppm")
```

Comment:
This part should probably be in the `hopper()` function not `_cached_hopper()`.

Reply:
With:
```python
def hopper(mode: str | None = None) -> Image.Image:
    # Use caching to reduce reading from disk but so an original copy is
    # returned each time and the cached image isn't modified by tests
    # (for fast, isolated, repeatable tests).
    if mode is None:
        # Always return fresh not-yet-loaded version of image.
        # Operations on not-yet-loaded images is separate class of errors
        # what we should catch.
        return Image.open("Tests/images/hopper.ppm")
    return _cached_hopper(mode).copy()
```

We get:
```python
============================================================================= ERRORS ==============================================================================
___________________________________________________________ ERROR collecting Tests/test_file_libtiff.py ___________________________________________________________
Tests/test_file_libtiff.py:53: in <module>
    class TestFileLibTiff(LibTiffTestCase):
Tests/test_file_libtiff.py:516: in TestFileLibTiff
    @pytest.mark.parametrize("im", (hopper("P"), Image.new("P", (1, 1), "#000")))
Tests/helper.py:263: in hopper
    return _cached_hopper(mode).copy()
Tests/helper.py:273: in _cached_hopper
    im = _cached_hopper().convert(mode)
Tests/helper.py:270: in _cached_hopper
    elif mode[:4] == "I;16":
E   TypeError: 'NoneType' object is not subscriptable
________________________________________________________ ERROR collecting Tests/test_file_libtiff_small.py ________________________________________________________
Tests/test_file_libtiff_small.py:8: in <module>
    from .test_file_libtiff import LibTiffTestCase
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:178: in exec_module
    exec(co, module.__dict__)
Tests/test_file_libtiff.py:53: in <module>
    class TestFileLibTiff(LibTiffTestCase):
Tests/test_file_libtiff.py:516: in TestFileLibTiff
    @pytest.mark.parametrize("im", (hopper("P"), Image.new("P", (1, 1), "#000")))
Tests/helper.py:263: in hopper
    return _cached_hopper(mode).copy()
Tests/helper.py:273: in _cached_hopper
    im = _cached_hopper().convert(mode)
Tests/helper.py:270: in _cached_hopper
    elif mode[:4] == "I;16":
E   TypeError: 'NoneType' object is not subscriptable
_________________________________________________________ ERROR collecting Tests/test_image_transpose.py __________________________________________________________
Tests/test_image_transpose.py:12: in <module>
    mode: helper.hopper(mode).crop((0, 0, 121, 127)).copy()
Tests/helper.py:263: in hopper
    return _cached_hopper(mode).copy()
Tests/helper.py:273: in _cached_hopper
    im = _cached_hopper().convert(mode)
Tests/helper.py:270: in _cached_hopper
    elif mode[:4] == "I;16":
E   TypeError: 'NoneType' object is not subscriptable
===================================================================== short test summary info =====================================================================
ERROR Tests/test_file_libtiff.py - TypeError: 'NoneType' object is not subscriptable
ERROR Tests/test_file_libtiff_small.py - TypeError: 'NoneType' object is not subscriptable
ERROR Tests/test_image_transpose.py - TypeError: 'NoneType' object is not subscriptable
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================== 3 errors in 2.82s ========================================================================
```

Reply:
Maybe put it in both?
That way the cached version is used within `_cached_hopper(...)` and a new instance is created every time `hopper(None)` is called.

Reply:
Sure, but `copy()` calls `load()`, whereas the comment here is specifically talking about returning an image object that has not been `load()`ed yet.

Reply:
Got it, thanks!

So to have a single check, we need to make sure the call in `_cached_hopper`'s `else` block goes through the initial `hopper()` instead of `_cached_hopper` again.

And we might as well make the call in the `if` go through `hopper`: it's only an extra function call and will be cached for subsequent calls.

Please see the newest commit.

Reply:
Looks good; you can now remove `None` from the type hint for `_cached_hopper()` since the `None` part is handled in `hopper()`.